src/dtx/dtx_common.c

```diff
@@ -0,0 +1,632 @@
+ * \param conflict	[IN]	Hash of the dkey to be modified if applicable.
+ * \param dti_cos	[IN,OUT]The DTX array to be committed because of shared.
+ * \param dti_cos_count [IN,OUT]The @dti_cos array size.
+ * \param pm_ver	[IN]	Pool map version for the DTX.
+ * \param intent	[IN]	The intent of related modification.
+ * \param leader	[IN]	The target (to be modified) is leader or not.
+ * \param dth		[OUT]	Pointer to the DTX handle.
+ *
+ * \return			Zero on success, negative value if error.
+ */
+int
+dtx_begin(struct dtx_id *dti, daos_unit_oid_t *oid, daos_handle_t coh,
+	  daos_epoch_t epoch, uint64_t dkey_hash,
+	  struct dtx_conflict_entry *conflict, struct dtx_id *dti_cos,
+	  int dti_cos_count, uint32_t pm_ver, uint32_t intent, bool leader,
+	  struct dtx_handle **dthp)
+{
+	struct dtx_handle	*dth;
+
+	if (leader) {
+		/* XXX: For leader case, we need to find out the potential
+		 *	conflict DTXs in the CoS cache, and append them to
+		 *	the dispatched RPC to non-leaders. Then non-leader
+		 *	replicas can commit them before real modifications
+		 *	to avoid availability trouble.
+		 */
+		D_ASSERT(dti_cos == NULL);
+		D_ASSERT(dti_cos_count == 0);
+
+		dti_cos_count = vos_dtx_list_cos(coh, oid, dkey_hash,
+				intent == DAOS_INTENT_UPDATE ?
+				DCLT_PUNCH : DCLT_PUNCH | DCLT_UPDATE,
+				DTX_THRESHOLD_COUNT, &dti_cos);
+		if (dti_cos_count < 0)
+			return dti_cos_count;
+
+		if (dti_cos_count > 0 && dti_cos == NULL) {
+			/* There are too many conflict DTXs to be committed,
+			 * as to cannot be taken via the normal IO RPC. The
+			 * background dedicated DTXs batched commit ULT has
+			 * not committed them in time. Let's retry later.
+			 */
+			D_DEBUG(DB_TRACE, "Too many pontential conflict DTXs "
+				"for the given "DF_DTI", let's retry later.\n",
+				DP_DTI(dti));
+			return -DER_INPROGRESS;
+		}
+	}
+
+	D_ALLOC_PTR(dth);
+	if (dth == NULL) {
+		if (leader && dti_cos != NULL)
+			D_FREE(dti_cos);
+
+		return -DER_NOMEM;
+	}
+
+	dth->dth_xid = *dti;
+	dth->dth_oid = *oid;
+	dth->dth_coh = coh;
+	dth->dth_epoch = epoch;
+	D_INIT_LIST_HEAD(&dth->dth_shares);
+	dth->dth_handled_time = crt_hlc_get();
+	dth->dth_dkey_hash = dkey_hash;
+	dth->dth_ver = pm_ver;
+	dth->dth_intent = intent;
+	dth->dth_sync = 0;
+	dth->dth_leader = (leader ? 1 : 0);
+	dth->dth_non_rep = 0;
+	dth->dth_dti_cos = dti_cos;
+	dth->dth_dti_cos_count = dti_cos_count;
+	dth->dth_conflict = conflict;
+	dth->dth_ent = UMMID_NULL;
+
+	*dthp = dth;
+	D_DEBUG(DB_TRACE, "Start the DTX "DF_DTI
+		" ver %u, dkey %llu, dti_cos_count %d, intent %s, %s\n",
+		DP_DTI(&dth->dth_xid), dth->dth_ver,
+		(unsigned long long)dth->dth_dkey_hash, dti_cos_count,
+		dth->dth_intent == DAOS_INTENT_PUNCH ? "Punch" : "Update",
+		leader ? "leader" : "non-leader");
+
+	return 0;
+}
+
+int
+dtx_end(struct dtx_handle *dth, struct ds_cont_hdl *cont_hdl,
+	struct ds_cont *cont, int result)
+{
+	int	rc = 0;
+
+	if (result < 0) {
```

Comment:
can it be positive value? If no, I'd suggest to use "if (result != 0)"

Reply:
It depends on the vos_update/punch return value, not sure whether it can be positive or not.